LocationProvider/src/main/java/com/marslogger/locationprovider/LocationProvider.java

```diff
@@ -0,0 +1,74 @@
+package com.marslogger.locationprovider;
+
+import android.Manifest;
+import android.content.Context;
+import android.content.pm.PackageManager;
+import android.location.Location;
+import android.os.HandlerThread;
+
+import androidx.annotation.NonNull;
+import androidx.core.app.ActivityCompat;
+import androidx.core.util.Consumer;
+
+import com.google.android.gms.location.FusedLocationProviderClient;
+import com.google.android.gms.location.LocationCallback;
+import com.google.android.gms.location.LocationRequest;
+import com.google.android.gms.location.LocationResult;
+import com.google.android.gms.location.LocationServices;
+
+public class LocationProvider {
+
+    private final Context context;
+
+    // background thread that listens for gps changes
+    private HandlerThread locationThread;
+    LocationCallback locationCallback;
+    FusedLocationProviderClient fusedLocationClient;
+
+    // for handling gps updates
+    private static final int DEFAULT_UPDATE_INTERVAL = 1;
+    private static final int FAST_UPDATE_INTERVAL = 1;
+
+    public LocationProvider(Context context) {
+        this.context = context;
+    }
+
+    public void createLocationListener(Consumer<Location> locationConsumer) {
+        locationThread = new HandlerThread("locationThread");
+        locationThread.start();
+        LocationRequest locationRequest = LocationRequest.create();
+        locationRequest.setInterval(500);
+        locationRequest.setFastestInterval(500);
+        locationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
+
+        fusedLocationClient = LocationServices.getFusedLocationProviderClient(context);
+
+        // check if you have permission before trying to get the location
+        if (ActivityCompat.checkSelfPermission(
+                context,
+                Manifest.permission.ACCESS_FINE_LOCATION
+        ) == PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(
+                context,
+                Manifest.permission.ACCESS_COARSE_LOCATION
+        ) == PackageManager.PERMISSION_GRANTED) {
+            fusedLocationClient.getLastLocation().addOnSuccessListener(locationConsumer::accept);
+        }
+
+        // triggered whenever the location is updated
+        locationCallback = new LocationCallback() {
+            @Override
+            public void onLocationResult(@NonNull LocationResult locationResult) {
+                super.onLocationResult(locationResult);
+                locationConsumer.accept(locationResult.getLastLocation());
```

Comment:
I think for saving location, my idea was to capture the location for each image only and for each imu data point associated with that image. Actually, for our own research project, we needed to capture the GPS value associated to each IMU data point. 

I do not understand your point about locations arriving in a batch. I don't think location values arrive in a batch.

Reply:
By a batch, I mean several location messages may arrive together, forming a batch. This will become common when the location update interval is long say 10 seconds.

Reply:
I guess I misunderstand how `FusedLocationProviderClient` works? Let me explain how I think it works and then you can correct me. 

I think if we set the location update interval to 10 seconds, that will mean that we check for a new location every 10 seconds. So, when we check for a new location after 10 seconds, there will be only one new location and not several locations. So, for example, we first set the `LocationInterval` object for 10 seconds:

```
LocationRequest locationRequest = LocationRequest.create();
locationRequest.setInterval(10000);
locationRequest.setFastestInterval(10000);
locationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);

// after setting up the FusedLocationProvider

// my understanding is that this line will get triggered after every 10 seconds on the locationThread
// hence, the locationCallback function will be triggered every 10 seconds and we will only get one new location value after 10 seconds
// am I wrong as to how this works?
fusedLocationClient.requestLocationUpdates(
        locationRequest, locationCallback, locationThread.getLooper()
);
```

Reply:
@JzHuai0108 is my understanding correct regarding FusedLocationProviderClient? Please let me know and I will make the necessary changes and update this pull request with the newly changed code

Reply:
Please take a look at [here](https://google-developer-training.github.io/android-developer-advanced-course-practicals/unit-4-add-geo-features-to-your-apps/lesson-7-location/7-1-p-use-the-device-location/7-1-p-use-the-device-location.html).
The quote "The fused location provider attempts to make location requests more efficient by batching requests from different apps. This means that you may receive updates faster than what you set in setInterval(), which can cause problems if your UI is not ready for updates. "
Another minor point worth mentioning is that the LocationResult contains a list of Locations, as described [here](https://developers.google.com/android/reference/com/google/android/gms/location/LocationResult).
Therefore, I think there will cases when a list of locations are available to the LocationCallback and we may be interested in save all these locations, see [here](https://droidbyme.medium.com/get-current-location-using-fusedlocationproviderclient-in-android-cb7ebf5ab88e).
After all, getLastLocation() is good enough.

Reply:
Thank you very much I will update the code by the end of this week.